### ISSUE TITLE
Template-able signal handler

### ISSUE THREAD

**Issue by rozap** [bitwalker/distillery#255]
Hi,

Distillery has been working really well for us so thank you! However, we're having issues with signal handlers. The `bootloader.eex` file appears to only allow for `rpc init stop` to be executed on a `TERM` signal. 

We're in a situation where we don't want the init system to take care of stopping our system because we want to have processes run to completion (ie: still be able to handle some messages after a signal, where the GenServer `stop` callback is really designed for final cleanup)

Basically what I'm getting at is: would it be reasonable to have an overlay var called something like `signal_handler` which could take on something like `rpc Elixir.OurApp.Shutdown shutdown` or is it possible to override `boot_loader.eex` entirely with our own which works differently?

Would this be a reasonable thing, or do you see potential pitfalls? If it seems reasonable I can try and make it work.

**Comment by bitwalker** [comment]
I'm open to the idea, it should be fairly easy to implement. If you want to take a stab at implementing it, we can work over the final details in the PR!

**Comment by michalmuskala** [comment]
One thing to notice in this context is the signal handlers that will be introduced in OTP 20 https://github.com/erlang/otp/pull/1315, so you'll be able to subscribe to any signal from inside the VM.

Also since OTP 19.3 `SIGTERM` is natively translated to `:init.stop()` by the runtime system.

**Comment by bitwalker** [comment]
Yeah, right now distillery intercepts signals to handle them better for legacy OTP versions, but I do plan to migrate to the let the runtime handle them again once they've been in the wild for a bit. We could also conditionally template parts of the boot script to let users choose to use the new signal handler in favor of the current behavior.

**Comment by michalmuskala** [comment]
I guess it should be possible to emulate the new OTP implementation on older versions using the wrapper script and a similar gen_event server that would be notified using rpc from the wrapper. Maybe that could be a way?

**Comment by rozap** [comment]
We actually installed a port driver which traps the signals in the VM. https://github.com/jkingsbery/sighandler 

It works well, but as @bitwalker said, there's a wrapper script which traps signals and does `rpc init stop`

I think with OTP 20, my preference would be to not intercept signals at all in that wrapper script and just let them get to the VM. IMO it's the least confusing option as you don't have to learn about the different layers of the release build. It should Just Work. 

I have an implementation at home that adds the custom rpc bit to the signal, I'll open a PR soon.

**Comment by DerKastellan** [comment]
I wanted to ask about an issue similar to the original one:

When I run my program with the "foreground" option and then Ctrl-C to end it, I observe that the associated beam.smp process is not actually killed/terminated, but the foreground process ends.

I had a look at https://github.com/bitwalker/distillery/blob/e6d6ea2af41238e1f396e01752098577801f5872/priv/templates/boot_loader.eex

The program is started normally in the main case because I made a printout.
On Ctrl-C the _cleanup is indeed called.

```
    if "$RELEASES_DIR/$REL_VSN/$REL_NAME.sh" rpc init stop >/dev/null; then
        wait $__pid
        exit $?
    else
        exit 0
    fi
```

It seems that the script passes into the else-branch because a printout before `wait` is not made.

However, the application still hangs around in the background.

I'm on OTP19. No funky things being done and no signal handling attempted in the application.

I could swear it worked last week.

**Comment by bitwalker** [comment]
As of master, releases on OTP 20 will use the new OTP signal handler, so I'm going to close this. Thanks for the discussion!
